res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1336 @@
+
+    TraktorS2MK2.controller.processButton(field);
+  }
+}
+
+// There are no buttons handled by the long message, so this is a little simpler.  (Even though
+// this is very similar to the other handler, it's easier to keep them separate to know what's
+// a control and what's a button.
+TraktorS2MK2.longMessageCallback = function(packet, data) {
+  for (name in data) {
+    field = data[name];
+    TraktorS2MK2.controller.processControl(field);
+  }
+}
+
+TraktorS2MK2.samplerGainKnob = function(field) {
+  for (var i = 1; i <= 8; i++) {
+    engine.setParameter("[Sampler" + i + "]", "pregain", field.value / 4096);
+  }
+}
+
+TraktorS2MK2.toggleButton = function(field) {
+  if (field.value > 0) {
+    script.toggleControl(field.group, field.name);
+  }
+}
+
+TraktorS2MK2.shift = function(field) {
+  var group = field.id.split(".")[0];
+  TraktorS2MK2.shift_pressed[group] = field.value;
+  TraktorS2MK2.outputCallback(field.value, field.group, "!shift");
+}
+
+TraktorS2MK2.loadTrack = function(field) {
+  var splitted = field.id.split(".");
+  var group = splitted[0];
+  if (TraktorS2MK2.shift_pressed[group]) {
+    engine.setValue(field.group, "eject", field.value);
+  } else {
+    engine.setValue(field.group, "LoadSelectedTrack", field.value);
+  }
+}
+
+TraktorS2MK2.syncEnabled = function(field) {
+  var now = Date.now();
+
+  var splitted = field.id.split(".");
+  var group = splitted[0];
+  // If shifted, just toggle.
+  // TODO(later version): actually make this enable explicit master.
+  if (TraktorS2MK2.shift_pressed[group]) {
+    if (field.value === 0) {
+      return;
+    }
+    var synced = engine.getValue(field.group, "sync_enabled");
+    engine.setValue(field.group, "sync_enabled", !synced);
+  } else {
+    if (field.value === 1) {
+      TraktorS2MK2.sync_enabled_time[field.group] = now;
+      engine.setValue(field.group, "sync_enabled", 1);
+    } else {
+      var cur_enabled = engine.getValue(field.group, "sync_enabled");
+      if (!cur_enabled) {
+        // If disabled, and switching to disable... stay disabled.
+        engine.setValue(field.group, "sync_enabled", 0);
+        return;
+      }
+      // was enabled, and button has been let go.  maybe latch it.
+      if (now - TraktorS2MK2.sync_enabled_time[field.group] > 300) {
+        engine.setValue(field.group, "sync_enabled", 1);
+        return;
+      }
+      engine.setValue(field.group, "sync_enabled", 0);
+    }
+  }
+}
+
+TraktorS2MK2.cue = function(field) {
+  var splitted = field.id.split(".");
+  var group = splitted[0];
+  if (TraktorS2MK2.shift_pressed[group]) {
+    if (ShiftCueButtonAction == "REWIND") {
+      if (field.value === 0) {
+        return;
+      }
+      engine.setValue(field.group, "start_stop", 1);
+    } else if (ShiftCueButtonAction == "REVERSEROLL") {
+      engine.setValue(field.group, "reverseroll", field.value);
+    } else {
+      print ("Traktor S2 Mk2 WARNING: Invalid ShiftCueButtonAction picked.  Must be either REWIND " +
```

Comment:
this should be checked on init and throw an error when the script is loaded.

Reply:
It would be easy to miss the warning that way. As it is, it is easy to find the warning in the log when the button is pressed.

Reply:
but it would pop up the alert right away -- I don't get how that would be missed.

Reply:
ah, I forgot we can now throw errors and have them show a popup

Reply:
done in 3df3b3a